# <font color="steelblue">Módulo Pandas: Estructuras de datos y detección de valores perdidos</font>

**Material desarrollado por los [equipos de trabajo de IA4LEGOS](https://ia4legos.umh.es/)**


**Fecha última edición**: 10/05/2025

**Licencia**: <small><a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br /></small>

No olvides hacer una copia si deseas utilizarlo. Al usar estos contenidos, aceptas nuestros términos de uso y nuestra política de privacidad.

In [ ]:
#@title <font color="steelblue" size="+1"> Descarga de ficheros de datos { display-mode: "form" }

# Descarga de ficheros
!wget {f"https://raw.githubusercontent.com/ia4legos/Statistics/main/data/corona.csv"}
!wget {f"https://raw.githubusercontent.com/ia4legos/Statistics/main/data/diabetes.csv"}
!wget {f"https://raw.githubusercontent.com/ia4legos/Statistics/main/data/encuesta_nutricion_cv.csv"}
!wget {f"https://raw.githubusercontent.com/ia4legos/Statistics/main/data/encuesta_nutricion_cv.xlsx"}

--2025-10-16 07:06:36--  https://raw.githubusercontent.com/ia4legos/Statistics/main/data/corona.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3879 (3.8K) [text/plain]
Saving to: ‘corona.csv’

corona.csv          100%[===================>]   3.79K  --.-KB/s    in 0s      

2025-10-16 07:06:36 (40.0 MB/s) - ‘corona.csv’ saved [3879/3879]

--2025-10-16 07:06:36--  https://raw.githubusercontent.com/ia4legos/Statistics/main/data/diabetes.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25822 (25K) [text/plain]
Saving to: ‘diab

## <font color='steelblue'>Introducción</font>

**Descripción:** Este cuaderno introduce el módulo *Pandas*, que es básico para la manipulación eficiente de bases de datos. Estudiamos dos tipos de estructuras de datos básicas como son las series y los dataframes. Aprendemos a crearlos de distintos modos, explorarlos, describirlos, seleccionar sus elementos y realizar operaciones con ellos. Introducimos también el concepto de valor perdido o faltante en una base de datos y algunas funciones para detectarlos.  Profundizamos en este cuaderno sobre la manipulación de dataframes de *Pandas* para llevar a cabo diversas ejecuciones comunes en la práctica estadística. Aprendemos a indexar un dataframe con información existente, y a ordenar, filtrar y agrupar variables con diversos criterios. Descubrimos funciones específicas para realizar cálculos más o menos complejos con pocas o muchas variables de un dataframe y también para categorizar variables numéricas y recodificar variables categóricas. Trabajamos algunas funcionalidades específicas de las series temporales. Por último presentamos el concepto de dato anómalo, identificamos sus efectos y presentamos métodos para reconocerlos y tratarlos antes de iniciar un análisis estadístico.

## <font color='steelblue'> Objetivos de aprendizaje

* Descubrir los tipos básicos de estructuras de datos en *Pandas*: series y dataframes.

* Crear y manipular series de datos de *Pandas* para trabajar con datos unidimensionales.

* Crear y manipular dataframes de *Pandas* para trabajar con bases de datos multidimensionales.

* Conocer el concepto de valor perdido o faltante en una base de datos, y aplicar funciones para reconocerlos en dataframes.

* Dominar la extracción selectiva de datos de una base de datos, realizar consultas, ordenar y filtrar información y agrupar datos según criterios específicos.

* Crear variables a partir de funciones programadas basadas en diversas variables de la base de datos, categorizar variables numéricas y recodificar variables categóricas.


* Convertir variables categóricas a variables binarias (variables *dummy*).

* Conocer conceptos y operaciones básicas con series temporales.

* Conocer el concepto de valor anómalo y aplicar técnicas estadísticas para identificarlos en una base de datos.





## <font color='steelblue'> Contenidos </font>

1. El módulo Pandas
1. Series de datos
1. Dataframes
1. Valores perdidos y detección
1. Tratamiento básico de datos
1. Crear y codificar variables
1. Variables *dummy*
1. Series temporales
1. Valores anónmalos

## <font color='steelblue'> 1. El módulo Pandas

*Pandas* es una de las librerías de Python más utilizadas como herramienta de alto nivel para la manipulación y el **análisis de datos** o **ciencia de datos**. Además de implementar una tipología de objetos como las series y los dataframes, óptimas para el tratamiento analítico y estadístico de bases de datos, contiene funciones que facilitan y hacen más eficiente el análisis.

Para trabajar con este módulo, importamos el módulo con la abreviatura `pd`.
```
import pandas as pd
```
Puesto que en este cuaderno trabajamos también con _NumPy_, y _matplotlib_, realizamos a continuación toda la carga de módulos precisa.

In [ ]:
# Importamos Pandas
import pandas as pd
# Importamos el resto de librerías
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
# Configuración de las figuras de matplotlib
plt.rcParams['figure.figsize'] = (15, 5)

### <font color='steelblue'>Estructuras de datos en Pandas

El módulo de **Python** llamado _Pandas_ proporciona unas estructuras de datos rápidas, flexibles y expresivas, diseñadas para que el trabajo con datos _relacionales_ o _etiquetados_ sea fácil e intuitivo.

_Pandas_ es adecuado para el tratamiento de diferentes tipos de datos:

- Datos tabulares con columnas de tipo heterogéneo, como en una tabla de SQL o una hoja de cálculo de Excel.
- Datos de series temporales, ordenados y no ordenados (no necesariamente de frecuencia fija).
- Matriz de datos arbitraria, con elementos homogéneos o  heterogéneos, y con etiquetas de filas y/o columnas.
- Cualquier otra forma de conjuntos de datos observacionales / estadísticos.

Este módulo nos ofrece las siguientes estructuras de datos:

* __Series__: se trata de _arrays_ unidimensionales indexados (con índice o etiqueta), similares a los diccionarios. Se pueden generar a partir de diccionarios o de listas.

* __DataFrame__: son estructuras de datos similares a las tablas de bases de datos relacionales (como SQL, o más sencillo Excel). También se pueden crear a partir de listas o diccionarios.

* __Panel, Panel4D y PanelND__: Estas estructuras de datos permiten trabajar con más de dos dimensiones. Dado que son estructuras complejas no trataremos los paneles en estos cuadernos.

Cuando tenemos un objeto definido con *Pandas*, para conocer de qué tipo (de los anteriores) es, utilizamos el comando:
```
**type(objeto)**
```

Algunas de las **ventajas** de usar _Pandas_ son:

- Facilita el manejo de grandes bases de datos, con datos de diferentes tipos, y en particular de los _datos faltantes_.

- Permite modificar de modo sencillo el tamaño de los objetos, como _insertar y eliminar_ columnas en un _DataFrame_, o realizar recortes y extraer subconjuntos de datos.

- Permite asignar de modo sencillo etiquetas a conjuntos de datos, automática o explícitamente.

- Proporciona herramientas robustas para cargar distintos tipos de archivos de datos (CSV, EXCEL, bases de datos, e incluso _HDF5 ultrarrápido_).

- Dispone de una gran cantidad de funciones para manipular datos agrupadas por funcionalidad (limpiar, transformar, visualizar,...).

- Sus funciones son muy eficientes para la manipulación de datos, y se integran fácilmente con otras librerías de Python.

- Tiene múltiples funcionalidades para tratamiento de datos de series temporales (fechas, frecuencias, ventanas temporales, ...).

- Permite transformar datos no estructurados para convertirlos en datos procesables. Los datos no estructurados son aquellos que no tienen una estructura interna identificable y por tanto carecen de valor hasta que se organizan (como por ejemplo las publicaciones en redes sociales).

Como desventajas cabe mencionar que tiene una curva de aprendizaje pronunciada al inicio, y que puede requerir de mucha memoria en el computador cuando se manipulan grandes bases de datos.

En conclusión, el módulo _Pandas_ es la herramienta base y más popular para realizar todas las tareas asumidas por los científicos de datos, desde el muestreo y la limpieza de datos, pasando por el análisis o modelado, hasta la visualización y la presentación de resultados.

### <font color='steelblue'>Tipos de variables</font>

En el ámbito estadístico hablamos de **variables** cuando tenemos que referirnos a características que observamos en la población de interés con algún fin informativo, explicativo o incluso predictivo. Así, para describir o explicar el gasto de las familias en nuestro país, podríamos recopilar el número de coches que tiene cada familia, los ingresos mensuales, cuánto se gasta cada semana a lo largo del año en la cesta de la compra, e incluso registrar qué productos componen dicha cesta. Todas las características que registramos para un estudio son variables y, como en el ejemplo, pueden ser de distintos tipos, hecho que influye en cómo las almacenamos y operamos con ellas.

Así pues, podemos hablar de distintos **tipos de datos**, o de **variables** que _Python_ identifica de modos distintos, y que se muestran en la tabla a continuación para los ejemplos que hemos planteado anteriormente:

| Tipo de dato/variable | Identificador  | Ejemplo |
| :-------------------- | :------------- | :------------- |
| **Cadena de texto** | `object (O)` | productos de la cesta de la compra |
| **Número entero** | `int64`    | número de coches en la familia |
| **Número decimal** | `float64` | ingresos mensuales |
| **Series temporales** | `datetime64` o `timedelta` | gasto semanal en la cesta de la compra |

El número '64' en los identificadores de enteros, decimales y series temporales, indica el número de bits de memoria que se asignan para almacenar datos en una celda.

Cuando leemos una base de datos recopilados sobre una muestra, a veces las variables registradas no tienen bien identificado el tipo de dato que almacena, por lo que siempre será preciso una revisión exhaustiva. Para ello existen varios métodos:

* **`.dtype`**, que aplicado a una variable, devuelve el identificador del tipo de dato que contiene.

Cuando agrupamos varias variables, por columnas, en un _dataframe_, podemos preguntar globalmente por el tipo de datos en todas las columnas con:

* **`.dtypes`**, que aplicado a un dataframe devuelve el tipo de dato para cada una de las variables/columnas.




En Estadística además, siempre hacemos otra clasificación (alternativa) de las variables, con la que tratamos de detallar, básicamente, cómo es la información que hemos recopilado, esto es, si se refiere a números, o a cualidades. Hablamos entonces de una diferenciación de las variables según dos grandes grupos, que a su vez se diferencia en otros subgrupos:

* **Variables cuantitativas:** cuyas observaciones registramos con números que tienen sentido como números, y que a su vez pueden ser:
  * __cuantitativas discretas__, que se expresan con números enteros (como el número de hijos, el número de trasplantes, el número de colegios en la comarca, ...)
  * __cuantitativas continuas__: números que aceptan decimales (como el bmi, el peso, la distancia entre hospitales, ...).

* **Variables cualitativas**, que nos propocionan información sobre una cualidad, una percepción, etc., esto es, no se responden/observan con un número, y nos permiten clasificar a los sujetos que estudiamos en grupos. Son variables de este tipo el color de ojos, la localidad de residencia, las respuestas a una encuesta de satisfacción, etc. Estos datos pueden ser cadenas de texto, pero también son representables mediante números enteros, y a veces con series temporales. Se distinguen dos subgrupos:

  * __cualitativas ordinales__, en las que las respuestas o valores observables posibles tienen un orden intrínseco, como en las encuestas de satisfacción las respuestas desde Muy insatisfecho a Muy satisfecho, y que a veces se plantean con una pregunta del tipo *'Puntúa del 1 al 5'*  para obtener ya directamente la respuesta codificada y ahorrar trabajo de codificación.
  
  * __cualitativas nominales__, en las que la respuesta es necesariamente una cadena de texto, como el color de ojos, o la localidad de residencia.

Las variables cualitativas son básicamente las variables de cadena de texto en Python, si bien para trabajar con ellas es habitual **codificar** las respuestas posibles con números que se utilizan como etiquetas para simplificar las respuestas posibles y facilitar el análisis. Así, el color de ojos se puede codificar, por ejemplo, con 1 para el verde, 2 para el azul, 3 para el marrón, y 4 para el resto, y esta codificación es arbitraria, generalmente decidida por el analista.

A partir de estas definiciones, encontraremos casos en los que variables cuantitativas discretas las tratemos como cuantitativas continuas (por ejemplo cuando calculamos el número medio de hijos en España), o incluso como cualitativas nominales (si nos interesa el porcentaje de sujetos que no tienen hijos frente a los que tienen al menos uno).
  
Una mención especial merecen las __variables cualitativas binarias o dicotómicas__, que son aquellas en las que exclusivamente se registran dos respuestas (cualitativas) posibles, como por ejemplo si un paciente es diabético o no. En estos tipos de variables se suele optar por una codificación 0/1.

Generalmente, cuando accedemos a una base de datos recopilada y almacenada por otros, se suele proporcionar una breve explicación del significado, respuestas posibles y la codificación de cada una de las variables observadas. Esta información nos permitirá asegurarnos de si efectivamente los tipos de datos que reconoce Python han sido bien registrados, o si no ha sido así, abordar las modificaciones pertinentes antes de iniciar el trabajo, para no errar en el análisis.

## <font color='steelblue'> 2. Dataframes

Un _dataframe_ de la librería _Pandas_ es el tipo de objeto útil para representar una base de datos con una estructura de matriz bidimensional, en la que las columnas representan variables y en cada fila se distribuyen los registros de todas las variables en cada uno de los sujetos u objetos observados o medidos. Estas filas, como en las series, están indexadas y los índices pueden ser sustituidos por etiquetas. Básicamente viene a representar una agrupación de series por columnas, con lo cual puede contener columnas de distintos tipos (texto, enteros, decimales,...), pero cada columna será de un único tipo. Por supuesto, todas las series agrupadas en un _dataframe_, esto es, todas las columnas, han de tener la misma longitud o número de elementos.

### <font color='steelblue'>Crear  </font>



Podemos crear _dataframes_ a partir de listas, arrays, y también a partir de diccionarios, con la función **`DataFrame()`** de *Pandas*. Además, podemos cargar como _dataframes_ archivos en diferentes formatos, disponibles en nuestro espacio de trabajo o en internet.



####<font color='steelblue'> Dataframes con listas

Para **crear _dataframes_ a partir de listas** hemos de:
1. agrupar las listas que representan las columnas en un array bidimensional, con **`np.stack()`**
1. definir dicha agrupación como dataframe con  **`pd.DataFrame()`**.

Vamos a generar una base de datos con la temperatura media diaria, durante una semana, de las capitales de provincia Alicante, Castellón y Valencia. Seguimos los pasos anteriores:

In [ ]:
# Definimos como listas los valores/registros de temperatura de cada ciudad
Alicante = [23.3, 19.8, 16.3, 14, 13.5, 13.4, 12.6]
Castellon = [20.9, 15.7, 15.3, 12.6, 13.3, 14.3, 14.1]
Valencia = [21.3, 15.3, 13.7, 10.7, 10.8, 8.5, 11.9]

#agrupamos por columnas las listas definidas
temperaturas = np.stack([Alicante, Castellon, Valencia], axis=1)

# definimos dicha agrupación como DataFrame
temperatura = pd.DataFrame(temperaturas)
print(temperatura)

      0     1     2
0  23.3  20.9  21.3
1  19.8  15.7  15.3
2  16.3  15.3  13.7
3  14.0  12.6  10.7
4  13.5  13.3  10.8
5  13.4  14.3   8.5
6  12.6  14.1  11.9


Comprobamos que la primera columna contiene los datos registrados para Alicante, la segunda de Castellón y la tercera de Valencia.

Si no indicamos nombres para las columnas, estas quedan  nombradas con números enteros correlativos que empiezan en 0. Podemos sin embargo, dar un nombre a cada columna asignando una lista (de nombres) al argumento `columns` de la función `DataFrame()`.

Repetimos la definición del dataframe, asignando nombres a las columnas a través del argumento **`columns^** en la función `DataFrame()`.

In [ ]:
# Creamos el dataframe de nuevo, indicando un nombre para las columnas
temperatura = pd.DataFrame(temperaturas, columns=["Alicante", "Castellón", "Valencia"])
print(temperatura)

   Alicante  Castellón  Valencia
0      23.3       20.9      21.3
1      19.8       15.7      15.3
2      16.3       15.3      13.7
3      14.0       12.6      10.7
4      13.5       13.3      10.8
5      13.4       14.3       8.5
6      12.6       14.1      11.9


Por defecto, las filas quedan indexadas numéricamente con la secuencia de enteros que inician en 0:
<center><img src=https://raw.githubusercontent.com/ia4legos/Statistics/main/images/img_160_30_10.png width="350" height="325"></center>



Al igual que podemos incluir los nombres de las columnas al crear el dataframe, podemos definir los nombres para los índices con el argumento **`index`** en `pd.DataFrame()`.

Definimos a continuación una lista con las etiquetas para los índices, y los incorporamos al crear el _dataframe_.


In [ ]:
indices = ["lunes","martes","miércoles","jueves","viernes","sábado","domingo"]

# definimos dicha agrupación como DataFrame
temperatura = pd.DataFrame(temperaturas,
                          columns=["Alicante", "Castellón", "Valencia"],
                          index = indices)
print(temperatura)

           Alicante  Castellón  Valencia
lunes          23.3       20.9      21.3
martes         19.8       15.7      15.3
miércoles      16.3       15.3      13.7
jueves         14.0       12.6      10.7
viernes        13.5       13.3      10.8
sábado         13.4       14.3       8.5
domingo        12.6       14.1      11.9


####<font color='steelblue'> Dataframes con diccionarios

También podemos **crear un _dataframe_ a partir de un diccionario**, resultando incluso más sencillo que con listas. Para ello basta con:
1. definir un diccionario donde la clave (_key_) es el nombre de la variable/columna en el _dataframe_, y los valores registrados vienen dados en una lista;
2. transformar el diccionario en un dataframe con la función `pd.DataFrame()`.

Como ejemplo vamos a reproducir el _dataframe_ anterior a través de un diccionario.

In [ ]:
# definimos un diccionario
temperaturas = {"Alicante": [23.3, 19.8, 16.3, 14, 13.5, 13.4, 12.6],
                "Castellon": [20.9, 15.7, 15.3, 12.6, 13.3, 14.3, 14.1],
                "Valencia": [21.3, 15.3, 13.7, 10.7, 10.8, 8.5, 11.9]}
indices = ["lunes","martes","miércoles","jueves","viernes","sábado","domingo"]

# transformamos el diccionario a dataframe
temperatura = pd.DataFrame(temperaturas,index = indices)
print(temperatura)

           Alicante  Castellon  Valencia
lunes          23.3       20.9      21.3
martes         19.8       15.7      15.3
miércoles      16.3       15.3      13.7
jueves         14.0       12.6      10.7
viernes        13.5       13.3      10.8
sábado         13.4       14.3       8.5
domingo        12.6       14.1      11.9


#### <font color='steelblue'>Dataframes desde archivos

Muchas veces las bases de datos con las que trabajamos proceden de otras fuentes y vienen dadas en ficheros de distintos tipos, con extensiones _csv_, _xlsx_, _html_, etc. Con _Pandas_ podemos leer fácilmente estos archivos y convertirlos directamente en _dataframes_ para poderlos manipular de forma eficiente. Disponemos de varias funciones en _Pandas_ para realizar la lectura en función del tipo de fichero/extensión:

* **`read_csv`**, lee ficheros con extensión _csv_ que se encuentran en nuestro espacio de trabajo, cuyos elementos  suelen ir delimitados por comas o tabuladores;
* **`read_table`**, lee ficheros _csv_ a los que accedemos mediante una url de acceso en internet;
* **`read_excel`**: para ficheros con extensión _xls_, _xlsx_, _xlsm_, _xlsb_, _odf_, _ods_ y _odt_;
* **`read_json`**: para ficheros en JavaScript con extensión _json_;
* **`read_html`**: lee tablas en formato _html_ y las transforma en listas o _dataframes_;
* **`read_sql`**: lee consultas o bases de datos de _SQL_.

Para ampliar información sobre el funcionamiento de estas funciones y consultar cómo leer otros tipos de archivos, te recomendamos consultar el [tutorial de Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html).




Vamos a ejemplificar cómo utilizar algunas de estas funciones, en concreto las que se utilizan con mayor frecuencia, como son las que permiten leer ficheros _csv_ en nuestro ordenador o espacio de trabajo (`read_csv`), y las que facilitan la lectura de archivos publicados y disponibles en internet (`read_table`).

Comenzamos ejemplificando el uso de `read_csv` con tres ficheros _csv_ que presentan distintas casuísticas y por lo tanto nos permitirán reflexionar. Estos ficheros han sido cargado en el directorio principal _/content_ al configurar este cuaderno y son:
- `corona.csv`,
- `diabetes.csv`  
- `encuesta_nutricion_cv.csv`.

Como siempre, si queremos manipular los datos que leemos con _Python_, hemos de asignar lo que leemos en un objeto de _Python_.

Recuerda que si necesitas ampliar información sobre esta función, es recomendable consultar el [tutorial de Pandas](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)

Comenzamos con la **base de datos 'corona.csv'**, que proporciona información sobre los casos de CoVid-19 en una fecha determinada. Podemos observar que esta base de datos tiene 199 filas y 3 columnas, es decir, 3 variables:

* _`country`_: variable cualitativa nominal (cadena de texto); identifica el país al que pertenecen los valores en cada fila;
* _`cases`_: variable cuantitativa continua (decimal positivo _float_); indica el número de casos positivos en covid registrados hasta la fecha;
* _`health_expenditure`_: variable cuantitativa continua (decimal positivo _float_); muestra el gasto (en millones de dólares) realizado en servicios sanitarios hasta la fecha.

Esta última variable contiene algunos valores etiquetados como _`NaN`_, correspondientes a valores perdidos (no se registró el dato y la celda quedó vacía). Este tipo de datos los trataremos más adelante en este cuaderno.


In [ ]:
# Lectura del fichero y creación de un dataframe
covid = pd.read_csv("corona.csv")
# comprobamos que la lectura es correcta
print(covid)
# comprobamos que es un dataframe
print(type(covid))
# comprobamos el tipo de las variables en el dataframe
print(covid.dtypes)

                    country    cases  health_expenditure
0                     China  81285.0               426.0
1                     Italy  74386.0              2700.0
2                       USA  65797.0                 NaN
3                     Spain  49515.0              2354.0
4                   Germany  37323.0              4592.0
..                      ...      ...                 ...
194        Papua New Guinea      1.0                77.0
195  St. Vincent Grenadines      1.0                 NaN
196                 Somalia      1.0                 NaN
197             Timor-Leste      1.0                72.0
198        Turks and Caicos      1.0                 NaN

[199 rows x 3 columns]
<class 'pandas.core.frame.DataFrame'>
country                object
cases                 float64
health_expenditure    float64
dtype: object


El segundo archivo, 'diabetes.csv', contiene información clínica de 768 pacientes respecto a 9 variables diferentes:

* _`pregnancies`_: variable cuantitativa discreta; indica el número de embarazos de la paciente
* _`glucose`_: variable cuantitativa discreta; proporciona los miligramos (_ml_ sin decimales) de glucosa en sangre cuando acudió a la consulta;
* _`diastolic`_: variable cuantitativa discreta; muestra la presión diastólica, medida en unidades enteras;
* _`triceps`_: variable cuantitativa discreta; es el perímetro del tríceps en centímetros (_cm_ sin decimales);
* _`insulin`_: variable cuantitativa discreta; son los miligramos de insulina en sangre (_mg_ sin decimales);
* _`bmi`_: variable cuantitativa continua; muestra el índice de masa corporal (con decimales);
* _`dpf`_: variable cuantitativa continua; _diabetes pedigree function_
* _`age`_: variable cuantitativa discreta; da la edad de la paciente;
* _`diabetes`_: variable cualitativa binaria o dicotómica, con 1 si tiene diabetes y 0 si no la tiene.



In [ ]:
# Lectura del fichero y creación de un dataframe
diabetes = pd.read_csv("diabetes.csv")
# comprobamos que la lectura es correcta
print(diabetes)
# comprobamos que es un dataframe
print(type(diabetes))
# comprobamos el tipo de las variables en el dataframe
print(diabetes.dtypes)

     pregnancies  glucose  diastolic  triceps  insulin   bmi    dpf  age  \
0              6      148         72       35        0  33.6  0.627   50   
1              1       85         66       29        0  26.6  0.351   31   
2              8      183         64        0        0  23.3  0.672   32   
3              1       89         66       23       94  28.1  0.167   21   
4              0      137         40       35      168  43.1  2.288   33   
..           ...      ...        ...      ...      ...   ...    ...  ...   
763           10      101         76       48      180  32.9  0.171   63   
764            2      122         70       27        0  36.8  0.340   27   
765            5      121         72       23      112  26.2  0.245   30   
766            1      126         60        0        0  30.1  0.349   47   
767            1       93         70       31        0  30.4  0.315   23   

     diabetes  
0           1  
1           0  
2           1  
3           0  
4      

Al visualizar un dataframe siempre obtenemos información sobre el tipo de cada una de las variables representadas en las columnas.

Si echas una ojeada a los archivos en el directorio de trabajo (con un doble click sobre el archivo), verás que tanto 'corona.csv' como 'diabetes.csv' se visualizan en nuestro cuaderno (en la ventana que se abre a la derecha) en formato de celdas, con las columnas (variables) perfectamente alineadas y separadas. Si abres con un editor de texto estos ficheros, verás que aparece cada registro en una fila, y los elementos de diferentes columnas, separados por una coma (`,`). Este es el formato que lee por defecto la función `read_csv`. No obstante, a veces los separadores son otros, y entonces el ajuste por defecto no proporciona una lectura óptima para poder trabajar con la base de datos. Hay que cambiar entonces los ajustes de lectura.

Prueba ahora a abrir y visualizar desde el directorio de trabajo el fichero `encuesta_nutricion_cv.csv`, y también prueba a leer como antes el fichero.

In [ ]:
# Lectura del fichero con los ajustes por defecto
encuesta = pd.read_csv("encuesta_nutricion_cv.csv")
# comprobamos qué se ha leído
print(encuesta)

      edad;peso;talla;sexo;provincia
0     18;66.4;163.5;HOMBRE;CASTELLÓN
1       18;68.7;172;HOMBRE;CASTELLÓN
2        24;76.7;170;HOMBRE;ALICANTE
3        17;55.6;160;MUJER;CASTELLÓN
4      16;46.6;163.5;MUJER;CASTELLÓN
...                              ...
1809      71;80.2;157;MUJER;VALENCIA
1810      81;68.3;150;MUJER;VALENCIA
1811    86;66.5;154.5;MUJER;VALENCIA
1812      67;74.1;150;MUJER;VALENCIA
1813      72;86.4;149;MUJER;VALENCIA

[1814 rows x 1 columns]


Al leer este archivo como los anteriores, comprobamos que todos los datos en una fila aparecen apiñados y separados por el punto y coma (`;`), de modo que no se diferencian variables ni valores: el dataframe no es operativo así. Para que la lectura sea correcta, debemos utilizar como argumento de `read_csv` el argumento **`sep=";"`**, que separa los valores de las diferentes variables en cada fila. Observa también que la primera línea contiene los nombres de las variables, separados por `;`.

Es importante pues, antes de leer un fichero, intentar visualizarlo para comprobar en qué formato viene, cómo se separan los valores y cuál es el carácter utilizado para separar las cifras decimales. Todas estas cuestiones las podemos ajustar como argumentos al llamar a la función `read_csv`.

Procedemos ya a cargar esta última base de datos, `encuesta_nutricion_cv.csv`, con los ajustes comentados.


In [ ]:
# Lectura del fichero con los ajustes adecuados
encuesta = pd.read_csv("encuesta_nutricion_cv.csv", sep= ";")
# comprobamos que la lectura es correcta
print(encuesta)
# comprobamos que es un dataframe
print(type(encuesta))
# comprobamos los tipos de las variables
encuesta.dtypes

      edad  peso  talla    sexo  provincia
0       18  66.4  163.5  HOMBRE  CASTELLÓN
1       18  68.7  172.0  HOMBRE  CASTELLÓN
2       24  76.7  170.0  HOMBRE   ALICANTE
3       17  55.6  160.0   MUJER  CASTELLÓN
4       16  46.6  163.5   MUJER  CASTELLÓN
...    ...   ...    ...     ...        ...
1809    71  80.2  157.0   MUJER   VALENCIA
1810    81  68.3  150.0   MUJER   VALENCIA
1811    86  66.5  154.5   MUJER   VALENCIA
1812    67  74.1  150.0   MUJER   VALENCIA
1813    72  86.4  149.0   MUJER   VALENCIA

[1814 rows x 5 columns]
<class 'pandas.core.frame.DataFrame'>


,0
edad,int64
peso,float64
talla,float64
sexo,object
provincia,object


Ahora sí se ha leído bien el _dataframe_.

Esta base de datos nos proporciona información sobre los datos de 1813 sujetos a los cuales se les ha hecho una encuesta nutricional en la Comunidad Valenciana. Se dispone de 5 variables que nos proporcionan información sobre ellos:

* _`edad`_: variable cuantitativa discreta, con valores enteros positivos que muestran la edad;
* _`peso`_: variable cuantitativa continua (decimales positivos) que indica el peso de cada persona en kilogramos;
* _`talla`_: variable cuantitativa continua(decimales) que indica la altura, en centímetros, de cada sujeto;
* _`sexo`_: variable cualitativa nominal (cadena de texto), que indica si el sujeto es hombre o mujer;
* _`provincia`_: variable cualitativa nominal (cadena de texto) que indica en qué provincia reside cada sujeto.

No siempre nos interesa trabajar con las variables tal cual vienen en la base de datos original. Así por ejemplo, nos podría interesar utilizar la variable _edad_ como cualitativa ordinal (no como cuantitativa), para agrupar a los sujetos por edad. Más tarde en este cuaderno veremos cómo modificar el tipo de las variables en los _dataframes_.

En ocasiones nos va a interesar **leer directamente un archivo de datos desde internet**, quizá desde algún repositorio web, sin necesidad de descargarlo en nuestro espacio de trabajo. Cuando dichos archivos están disponibles en la web como _dataframes_ (generalmente los repositorios proporcionan previsualizaciones con las que comprobarlo, o los ficheros tienen una extensión .TSV -valores separados tabularmente-), podemos leerlos directamente como _dataframes_ en nuestro cuaderno Colab con la función **`read_table`** del módulo _Pandas_.

Para ejemplificar el uso de esta función, utilizamos un archivo publicado en GitHub, llamado [`chipotle.tsv`](https://raw.githubusercontent.com/justmarkham/pandas-videos/master/data/chipotle.tsv)

In [ ]:
# Leemos el dataset de ordenes de Chipotle de una URL
# y guardamos los resultados en un dataframe
pedidos = pd.read_table('http://bit.ly/chiporders')

# comprobamos que la lectura es correcta
print(pedidos)
# comprobamos que es un dataframe
print(type(pedidos))

# comprobamos los tipos de las variables
print(pedidos.dtypes)


      order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]  

### <font color='steelblue'>Explorar
Cuando tenemos una base de datos, los primeros pasos a dar consisten en inspeccionarla, esto es, explorar cuál es su dimensión, cómo se presentan los registros y variables, e incluso acceder a los valores que contiene para ver a continuación si la lectura se ha realizado correctamente o hay algún error. En ocasiones nos interesará modificar los nombres de las columnas y/o de las filas. A veces tendremos que incorporar alguna variable más a modo de columna, e incluso eliminar alguna columna o algún registro.

Vemos cómo nos ayudan en estas tareas los métodos y funciones `shape`, `head`, `tail`, `index`, `columns`, `rename` y `drop`.



Cuando mostramos por pantalla un dataframe ya nos indica su **dimensión**, esto es, el número de filas y columnas. Si necesitamos dicha información en una tupla de números, tenemos el método `shape`.


In [ ]:
# tupla con la dimensión del dataframe
temperatura.shape

(7, 3)

Generalmente las bases de datos son muy voluminosas, por lo que es poco práctico visualizarlas en pantalla íntegramente. Con todo,siempre interesante ver cómo están estructuradas y cuáles son sus columnas. Para **visualizar solo unas cuantas filas del inicio o del final** tenemos las funciones **`head()`** y **`tail()`** respectivamente. Por defecto, con estas funciones, se muestran los cinco primeros (últimos) elementos, si bien se puede especificar cuántos mostrar (hasta un máximo de 60).

El dataframe 'encuesta' tiene un total de 1814 filas, por lo que es inviable visualizarlas todas en pantalla. Mostremos, con estas funciones, solo unas pocas al inicio y al final.

In [ ]:
# Mostramos las 5 primeras filas
print("\n Mostramos las 5 primeras filas \n",encuesta.head())

# Mostramos las 3 primeras filas
print("\n Mostramos las 3 primeras filas \n",encuesta.head(3))

# Mostramos las 5 últimas filas
print("\n Mostramos las 5 últimas filas \n",encuesta.tail())

# Mostramos las 3 primeras filas
print("\n Mostramos las 3 últimas filas \n",encuesta.tail(3))


 Mostramos las 5 primeras filas 
    edad  peso  talla    sexo  provincia
0    18  66.4  163.5  HOMBRE  CASTELLÓN
1    18  68.7  172.0  HOMBRE  CASTELLÓN
2    24  76.7  170.0  HOMBRE   ALICANTE
3    17  55.6  160.0   MUJER  CASTELLÓN
4    16  46.6  163.5   MUJER  CASTELLÓN

 Mostramos las 3 primeras filas 
    edad  peso  talla    sexo  provincia
0    18  66.4  163.5  HOMBRE  CASTELLÓN
1    18  68.7  172.0  HOMBRE  CASTELLÓN
2    24  76.7  170.0  HOMBRE   ALICANTE

 Mostramos las 5 últimas filas 
       edad  peso  talla   sexo provincia
1809    71  80.2  157.0  MUJER  VALENCIA
1810    81  68.3  150.0  MUJER  VALENCIA
1811    86  66.5  154.5  MUJER  VALENCIA
1812    67  74.1  150.0  MUJER  VALENCIA
1813    72  86.4  149.0  MUJER  VALENCIA

 Mostramos las 3 últimas filas 
       edad  peso  talla   sexo provincia
1811    86  66.5  154.5  MUJER  VALENCIA
1812    67  74.1  150.0  MUJER  VALENCIA
1813    72  86.4  149.0  MUJER  VALENCIA


Para saber **cuál es el nombre de los índices y de las columnas** de un _dataframe_ utilizamos el método `.index` y `.columns` respectivamente.

In [ ]:
# Nombres de filas (índices) y columnas (variables)
print("Índices: ", temperatura.index)
print("Columnas: ", temperatura.columns)

Índices:  Index(['lunes', 'martes', 'miércoles', 'jueves', 'viernes', 'sábado',
       'domingo'],
      dtype='object')
Columnas:  Index(['Alicante', 'Castellon', 'Valencia'], dtype='object')


Si queremos **modificar el nombre de las columnas o de los índices** en un dataframe tenemos dos opciones sencillas:

1. crear una lista con los nuevos nombres y asignarla a la lista que contiene las columnas/índices del dataframe,
```
dataframe.columns = lista_nombres
dataframe.index = lista_nombres
```

2. aplicar al dataframe la función **`.rename()`** con el argumento `columns` o `index` según corresponda, y especificando en un diccionario todas las modificaciones pertinentes `{'nombre_actual':'nombre_nuevo'}`. El argumento `inplace=True` indica que las actualizaciones sustituyen las antiguas variables; si no lo utilizamos, se asume `inplace=False`, y entonces habremos de asignar el cambio a un nuevo objeto para conservar el original.

Renombramos todas las columnas en 'temperatura' con los dos métodos propuestos.

In [ ]:
# Modificación de nombres de columnas

# Partimos del dataframe original
temperatura = pd.DataFrame(temperaturas, columns=["Alicante", "Castellón", "Valencia"])
print(temperatura)
# modificamos el nombre de las variables en 'temperatura'
temperatura.columns = ["Albacete", "Murcia", "Tarragona"]
print(temperatura)

# Recuperamos el nombre de las variables con 'rename'
temperatura.rename(columns={'Albacete':'Alicante', 'Murcia':'Castellón',
                            'Tarragona': 'Valencia'}, inplace=True)
print(temperatura)

   Alicante Castellón  Valencia
0      23.3       NaN      21.3
1      19.8       NaN      15.3
2      16.3       NaN      13.7
3      14.0       NaN      10.7
4      13.5       NaN      10.8
5      13.4       NaN       8.5
6      12.6       NaN      11.9
   Albacete Murcia  Tarragona
0      23.3    NaN       21.3
1      19.8    NaN       15.3
2      16.3    NaN       13.7
3      14.0    NaN       10.7
4      13.5    NaN       10.8
5      13.4    NaN        8.5
6      12.6    NaN       11.9
   Alicante Castellón  Valencia
0      23.3       NaN      21.3
1      19.8       NaN      15.3
2      16.3       NaN      13.7
3      14.0       NaN      10.7
4      13.5       NaN      10.8
5      13.4       NaN       8.5
6      12.6       NaN      11.9


Si una vez explorada la base de datos nos falta alguna información, a modo de nueva variable, podemos **añadir muy fácilmente una columna nueva**. Simplemente llamamos al dataframe con el operador de selección `[]`, e incluimos dentro de los paréntesis el nombre entrecomillado para la nueva columna. A esta selección le asignamos los valores que deseamos, organizados en una lista.
```
dataframe["nueva_variable"] = lista_valores
```

Al igual que a veces necesitamos incorporar nuevas columnas, en ocasiones interesa **eliminar filas y/o columnas** en un dataframe, para lo que disponemos de la función **`.drop()`** de *Pandas*. Esta función aplica el argumento `inplace=False`, que genera un nuevo dataframe para conservar el original, por lo que el cambio habremos de asignarlo a un nuevo objeto para conservarlo.

La eliminación de filas y columnas podemos hacerlas tanto con numerales como con los nombres o etiquetas correspondientes.

- para eliminar una o varias filas, incluidas en una lista con las etiquetas de las filas correspondientes:
  *  `dataframe.drop(lista_indices)`
  *  `dataframe.drop(lista_indices, axis=0)`
- para eliminar una o varias columnas, incluidas en una lista con los nombres de las columnas correspondientes:
  *  `dataframe.drop(columns=lista_variables)`
  *  `dataframe.drop(lista_variables, axis=1)`.

Si no recordamos las etiquetas o nombres de los índices o columnas, siempre podemos recuperarlos a partir de las listas que los contienen, `dataframe.index` y `dataframe.columns`.

Vemos a continuación cómo trabajamos estas funciones con el dataframe 'temperatura' para eliminar registros o filas.


In [ ]:
# Eliminar filas en un dataframe

# Partimos del dataframe original
temperatura = pd.DataFrame(temperaturas,
                           columns=["Alicante", "Castellón", "Valencia"])
# eliminamos las dos primeras filas
print(temperatura.drop(range(0,2)))
# eliminamos las filas 3 y 4
print(temperatura.drop([2,3],axis=0))

# renombramos índices y eliminamos el martes y el jueves
temperatura.index=indices
print(temperatura.drop(['martes','jueves']))

# eliminamos el primer registro pero no recordamos su etiqueta
print(temperatura.drop(temperatura.index[0]))


   Alicante Castellón  Valencia
2      16.3       NaN      13.7
3      14.0       NaN      10.7
4      13.5       NaN      10.8
5      13.4       NaN       8.5
6      12.6       NaN      11.9
   Alicante Castellón  Valencia
0      23.3       NaN      21.3
1      19.8       NaN      15.3
4      13.5       NaN      10.8
5      13.4       NaN       8.5
6      12.6       NaN      11.9
           Alicante Castellón  Valencia
lunes          23.3       NaN      21.3
miércoles      16.3       NaN      13.7
viernes        13.5       NaN      10.8
sábado         13.4       NaN       8.5
domingo        12.6       NaN      11.9
           Alicante Castellón  Valencia
martes         19.8       NaN      15.3
miércoles      16.3       NaN      13.7
jueves         14.0       NaN      10.7
viernes        13.5       NaN      10.8
sábado         13.4       NaN       8.5
domingo        12.6       NaN      11.9


Y ahora lo hacemos para eliminar columnas.

In [ ]:
# Eliminar columnas en un dataframe

# Eliminamos las dos primeras columnas con diferentes códigos
print(temperatura.drop(columns=["Alicante","Castellón"]))
print(temperatura.drop(["Alicante","Castellón"],axis=1))
print(temperatura.drop(temperatura.columns[0:2],axis=1))

           Valencia
lunes          21.3
martes         15.3
miércoles      13.7
jueves         10.7
viernes        10.8
sábado          8.5
domingo        11.9
           Valencia
lunes          21.3
martes         15.3
miércoles      13.7
jueves         10.7
viernes        10.8
sábado          8.5
domingo        11.9
           Valencia
lunes          21.3
martes         15.3
miércoles      13.7
jueves         10.7
viernes        10.8
sábado          8.5
domingo        11.9


Para saber más sobre la función `drop` se recomienda consultar el [Tutorial de Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html).

### <font color='steelblue'>Describir



Una vez explorada la base de datos, los siguientes pasos tienen que ver con la descripción de la misma, y en particular de las variables que contiene. Trabajamos en este apartado cómo conseguir información sobre el tipo de datos en cada una de las columnas, y también cómo modificarlos cuando nos interese hacer un cambio (funciones `info()`, `dtype()`, `dtypes()` y `astype()`).


Para obtener la información básica sobre una base de datos, esto es, formato de la base de datos, nombre de las columnas, número de datos no nulos  y el tipo de variable, podemos utilizar la función **`info()`**.

In [ ]:
# Lectura del fichero 'encuesta'
encuesta = pd.read_csv("encuesta_nutricion_cv.csv", sep= ";")
# información básica sobre la base de datos
encuesta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1814 entries, 0 to 1813
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   edad       1814 non-null   int64  
 1   peso       1777 non-null   float64
 2   talla      1777 non-null   float64
 3   sexo       1814 non-null   object 
 4   provincia  1814 non-null   object 
dtypes: float64(2), int64(1), object(2)
memory usage: 71.0+ KB


La función `info()` aplicada a la 'encuesta' nos indica que la base de datos en un dataframe de _Pandas_, con 1814 registros no nulos en todas las variables `edad`, `sexo` y `provincia`, y 1777 en `peso` y `talla`, y nos da el tipo de cada una de ellas: `edad` entera, peso y `talla` decimales, y `sexo` y `provincia` categóricas (cadena de caracteres); resume a continuación el número de variables de cada tipo, y da la cantidad de memoria que usa (más de 71KB).

Específicamente para **comprobar el tipo de las variables** en un dataframe  de un modo individual disponemos del método **`.dtype`**, y para conseguir el tipo de todas a la vez contamos con **`dtypes`**.

In [ ]:
# Mostramos todos los tipos de las variables
print(encuesta.dtypes)
# Mostramos el tipo de una sola variable
print("\nEdad es de tipo:", encuesta["edad"].dtype)

edad           int64
peso         float64
talla        float64
sexo          object
provincia     object
dtype: object

Edad es de tipo: int64


Así tenemos que la `edad` viene dada en números enteros, el `peso` y la `talla` en decimales, y el `sexo` y la `provincia` en cadenas de texto (*object*).

En ocasiones querremos **modificar el tipo de alguna de las variables** en un dataframe, pues no siempre encontraremos las bases de datos convenientemente procesadas y depuradas. También será común que en base a los objetivos de nuestro análisis necesitemos modificar el tipo de algunas variables. Para ello basta con utilizar la función **`astype()`** de Pandas, indicando dentro del paréntesis el tipo al que se desea cambiar.


Así por ejemplo, podríamos decidir trabajar la 'encuesta' con `age` como variable entera o float si nos interesa calcular medias, o como cualitativa ordinal si nos interesa agrupar por edades. Si queremos trabajar con la variable `edad` como cualitativa hemos de cambiar su tipo a `str`, lo que conseguimos al aplicar la función `astype(str)`.

In [ ]:
# Cambio de tipo de una variable entera a string, en un dataframe
encuesta["edad"] = encuesta["edad"].astype(str)
# comprobamos que se ha cambiado el tipo
encuesta['edad'].dtype

dtype('O')

**Para obtener las frecuencias absolutas o conteos en todas las categorías de respuesta posibles de una variable/columna categórica**, disponemos de la función **`value_counts()`**, que aplicamos a la variable seleccionada.

La probamos con la variable `provincia`.

In [ ]:
# obtenemos las frecuencias de respuesta en la variable 'provincia'
encuesta["provincia"].value_counts()

,count
provincia,
VALENCIA,925
ALICANTE,665
CASTELLÓN,224


### <font color='steelblue'>Seleccionar

Cuando trabajamos sobre una base de datos será frecuente la selección de elementos, filas, columnas o una combinación de ambas, para ir resolviendo cuestiones de interés. Aprendemos aquí los modos básicos de selección en dataframes, tanto con el operador de selección `[...]` como con los métodos `loc` e `iloc`.

Para seleccionar cierta columna de un _dataframe_ podemos utilizar directamente el nombre de variable que representa, entrecomillado, dentro de los operadores de selección `[]`, y aplicados al *dataframe*, **`dataframe['nombre_variable']`**.

Otra opción es escribir el nombre del dataframe, y aplicar a continuación el punto y el nombre de la variable, **`dataframe.nombre_variable`**.

Si se trata de seleccionar varias columnas, las especificamos seleccionando sobre el dataframe una lista con sus nombres entrecomillados, **`dataframe[["var1","var2",...]]`**.

Como ejemplo, seleccionamos las columnas con las temperaturas de Alicante y Valencia en el dataframe 'temperatura'.

In [ ]:
# Selección de una columna/variable
print("Selección de una columna")

print("\nModo1: dataframe[nombre_variable]")
print(temperatura["Alicante"])

print("\nModo2: dataframe.nombre_variable")
print(temperatura.Alicante)

# Selección de varias columnas/variables
print("\nSelección de varias columnas")
print(temperatura[["Alicante","Valencia"]])

Selección de una columna

Modo1: dataframe[nombre_variable]
lunes        23.3
martes       19.8
miércoles    16.3
jueves       14.0
viernes      13.5
sábado       13.4
domingo      12.6
Name: Alicante, dtype: float64

Modo2: dataframe.nombre_variable
lunes        23.3
martes       19.8
miércoles    16.3
jueves       14.0
viernes      13.5
sábado       13.4
domingo      12.6
Name: Alicante, dtype: float64

Selección de varias columnas
           Alicante  Valencia
lunes          23.3      21.3
martes         19.8      15.3
miércoles      16.3      13.7
jueves         14.0      10.7
viernes        13.5      10.8
sábado         13.4       8.5
domingo        12.6      11.9


Existe sin embargo un método más genérico para seleccionar filas y columnas, a través de dos funciones específicas como son:
- `iloc[]` para seleccionar filas y columnas utilizando los numerales que las identifican
- `loc[]` para seleccionar filas y columnas utilizando los nombres o etiquetas que las identifican.

En ambas, como argumentos podemos pasar una etiqueta única, una lista de etiquetas, una porción de etiquetas o una serie booleana (de datos `True/False`).

Estos métodos se aplican a un data frame seguidos de un punto, y con la selección entre corchetes:
```
dataframe.loc[seleccion]
dataframe.iloc[seleccion]
```

Las reglas de selección con estas funciones son las siguientes:
- seleccionar toda una fila *f*, `.iloc[f] / .loc[f]`,
- seleccionar toda una columna *c*, `.iloc[:,c] / .loc[:,c]`
- realizar múltiples selecciones de filas y columnas, `.iloc[lista_f,lista_c] / .loc[lista_f,lista_c]`.

Veamos algunos ejemplos, primero seleccionando por los numerales con `iloc` y luego con las etiquetas con `loc`.

In [ ]:
# Selección por numerales

# Selección de la primera fila
print("\n Primera fila:\n",encuesta.iloc[0])
# selección de la última fila
print("\n Última fila:\n",encuesta.iloc[-1])

# Selección de la primera columna
print("\n Primera columna:\n",encuesta.iloc[:,0])
# selección de la última columna
print("\n Última columna:\n",encuesta.iloc[:,-1])


 Primera fila:
 edad                18
peso              66.4
talla            163.5
sexo            HOMBRE
provincia    CASTELLÓN
Name: 0, dtype: object

 Última fila:
 edad               72
peso             86.4
talla           149.0
sexo            MUJER
provincia    VALENCIA
Name: 1813, dtype: object

 Primera columna:
 0       18
1       18
2       24
3       17
4       16
        ..
1809    71
1810    81
1811    86
1812    67
1813    72
Name: edad, Length: 1814, dtype: object

 Última columna:
 0       CASTELLÓN
1       CASTELLÓN
2        ALICANTE
3       CASTELLÓN
4       CASTELLÓN
          ...    
1809     VALENCIA
1810     VALENCIA
1811     VALENCIA
1812     VALENCIA
1813     VALENCIA
Name: provincia, Length: 1814, dtype: object


In [ ]:
# Selección de varias filas y columnas
print("\n Primeras dos filas y primeras dos columnas:\n",encuesta.iloc[0:2,0:2])

# selección de la primera, tercera y cuarta filas (todas las variables)
print("\n primera, tercera y cuarta filas:\n",encuesta.iloc[[0,2,3]])
print("\n primera, tercera y cuarta filas:\n",encuesta.iloc[[0,2,3],:])

# selección de la primera y tercera columnas
print("\n primera, tercera y cuarta columnas:\n",encuesta.iloc[:,[0,2]])



 Primeras dos filas y primeras dos columnas:
   edad  peso
0   18  66.4
1   18  68.7

 primera, tercera y cuarta filas:
   edad  peso  talla    sexo  provincia
0   18  66.4  163.5  HOMBRE  CASTELLÓN
2   24  76.7  170.0  HOMBRE   ALICANTE
3   17  55.6  160.0   MUJER  CASTELLÓN

 primera, tercera y cuarta filas:
   edad  peso  talla    sexo  provincia
0   18  66.4  163.5  HOMBRE  CASTELLÓN
2   24  76.7  170.0  HOMBRE   ALICANTE
3   17  55.6  160.0   MUJER  CASTELLÓN

 primera, tercera y cuarta columnas:
      edad  talla
0      18  163.5
1      18  172.0
2      24  170.0
3      17  160.0
4      16  163.5
...   ...    ...
1809   71  157.0
1810   81  150.0
1811   86  154.5
1812   67  150.0
1813   72  149.0

[1814 rows x 2 columns]


Si queremos seleccionar utilizando las etiquetas o nombres de las filas (índices) o de las columnas, en lugar de utilizar `iloc`, hemos de utilizar **`loc`**.

In [ ]:
# Seleccionar varias columnas
temperatura.loc[:,["Alicante","Valencia"]]

,Alicante,Valencia
lunes,23.3,21.3
martes,19.8,15.3
miércoles,16.3,13.7
jueves,14.0,10.7
viernes,13.5,10.8
sábado,13.4,8.5
domingo,12.6,11.9


In [ ]:
print(temperatura.index)
print(temperatura.index[2:4])

Index(['lunes', 'martes', 'miércoles', 'jueves', 'viernes', 'sábado',
       'domingo'],
      dtype='object')
Index(['miércoles', 'jueves'], dtype='object')


In [ ]:
# Seleccionar varias filas por etiquetas
temperatura.index
print("Seleccionar las filas de los lunes:", temperatura.loc[temperatura.index == 'lunes'])
print("\n")
print("Seleccionar dos filas correlativas a partir de la tercera:", temperatura.loc[temperatura.index[2:4]])

Seleccionar las filas de los lunes:        Alicante Castellón  Valencia
lunes      23.3       NaN      21.3


Seleccionar dos filas correlativas a partir de la tercera:            Alicante Castellón  Valencia
miércoles      16.3       NaN      13.7
jueves         14.0       NaN      10.7


Por último, si queremos **generar un *array* solo con los valores de una o varias columnas**, utilizamos el comando **`.values`**, aplicado a la selección de interés sobre el dataframe.

In [ ]:
# array con los valores de las temperaturas de Alicante
print("Temperaturas de Alicante")
temperatura["Alicante"].values


Temperaturas de Alicante


array([23.3, 19.8, 16.3, 14. , 13.5, 13.4, 12.6])

In [ ]:
print("Temperaturas de Alicante y Valencia")
temperatura[["Alicante","Valencia"]].values

Temperaturas de Alicante y Valencia


array([[23.3, 21.3],
       [19.8, 15.3],
       [16.3, 13.7],
       [14. , 10.7],
       [13.5, 10.8],
       [13.4,  8.5],
       [12.6, 11.9]])

In [ ]:
# Mostramos todos los valores de todas las columnas a modo de listas (aunque son arrays)
for columna in temperatura.columns:
  print(columna, temperatura[columna].values)

Alicante [23.3 19.8 16.3 14.  13.5 13.4 12.6]
Castellón [nan nan nan nan nan nan nan]
Valencia [21.3 15.3 13.7 10.7 10.8  8.5 11.9]


## <font color='steelblue'>3. Valores perdidos y detección

Con frecuencia, cuando leemos o incluso creamos una base de datos, podemos encontrar registros en los que no disponemos de información, bien porque no se ha observado/medido, bien porque se ha producido un error en la transcripción y simplemente no se registró. Hablamos aquí sobre  los valores perdidos en una base de datos, la relevancia que pueden tener en un análisis y cómo detectarlos.


Cuando pedimos información sobre la base de datos `encuesta`, ya comprobamos que las  variables _talla_ y _peso_ tenían solo 1777 registros, frente a las 1814 filas del fichero. Es decir, contenían valores perdidos, también llamados valores faltantes. Estos valores se identifican como `NaN` o `nan` en _Python_, que en inglés significa *Not A Number*, y se obtienen con `NaN` o con `np.nan`, del módulo  _NumPy_.

Cuando utilizamos el comando de *Pandas* `read_csv` para leer bases de datos, esta función detecta automáticamente los valores perdidos y los reemplaza con este valor especial `NaN`.

Los valores faltantes en una base de datos pueden llegar a ser relevantes y ocasionar sesgos importantes en las conclusiones, si están vinculados a muchos registros, o a registros con ciertas características. Es pues SIEMPRE muy importante, identificarlos al comenzar a trabajar con un conjunto de datos.

En ocasiones, los investigadores codifican de un modo concreto los valores perdidos (por ejemplo, es común encontrar el valor -1, cuando todos los valores han de ser necesariamente positivos en una variable). Si este es el caso, antes de iniciar el trabajo con la base de datos, habremos de modificar estos valores asignados a los registros perdidos, y asignar el valor `NaN` para poderlo trabajar convenientemente en Python como valor perdido, y excluir por tanto dichos registros en algunos análisis (por ejemplo, no tiene sentido integrar registros `NaN` para calcular un promedio).

Una vez reconocidos adecuadamente los valores faltantes en una base de datos, las estrategias básicas para el tratamiento de los datos son:

1. eliminar todos los registros con datos faltantes, y utilizar exclusivamente aquellos registros que contienen información completa en todas las variables;
2. otra opción, si hay muchos valores faltantes e información suficiente, es imputar valores para no desperdiciar tanta información excluyendo los registros perdidos.

Si bien no trabajaremos en este cuaderno estas estrategias (lo haremos más adelante), sí que presentamos a continuación las funciones que nos ayudarán a detectar valores perdidos
en una base de datos.



### <font color='steelblue'>Funciones para detectar valores perdidos

En los procedimientos básicos del análisis estadístico, *Python* omite este tipo de valores, pero en algunos casos más complejos es conveniente poder identificarlos para trabajarlos de forma adecuada.

Para detectar los valores perdidos, podemos utilizar las siguientes funciones:

* **`isnull()`**: proporciona valor `True` cuando una observación es faltante y `False` para las no faltantes;
* **`notnull()`**: procede al revés de la anterior, e identifica con `True` los valores que no son faltantes;
* **`dropna()`**: filtra los datos eliminando datos faltantes según el criterio que utilicemos como argumento; `how=any` excluya las filas en las que encuentra algún valor faltante en alguna columna; `how=all` excluye solo las filas en las que todas las columnas tienen valores faltantes; `subset='var1'` excluye solo las filas en las que la variable `var1` tiene valores faltantes.


Aplicamos a continuación estas funciones a la base de datos 'encuesta'.  

Al aplicar la función `info()`, observamos la presencia de valores faltantes en las variables `peso` y `talla`, con 1777 registros no nulos cada una, a diferencia de 1814 que es el número total de registros existentes.


In [ ]:
# Información básica sobre variables con valores faltantes
encuesta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1814 entries, 0 to 1813
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   edad       1814 non-null   object 
 1   peso       1777 non-null   float64
 2   talla      1777 non-null   float64
 3   sexo       1814 non-null   object 
 4   provincia  1814 non-null   object 
dtypes: float64(2), object(3)
memory usage: 71.0+ KB


Para identificar cuáles son los datos faltantes, usamos `isnull()`, si bien es poco esclareceror en la visualización si la base de datos es razonablemente grande.


In [ ]:
# Visualizamos la identificación de faltantes en el dataframe
encuesta.isnull()

,edad,peso,talla,sexo,provincia
0,False,False,False,False,False
1,False,False,False,False,False
2,False,False,False,False,False
3,False,False,False,False,False
4,False,False,False,False,False
...,...,...,...,...,...
1809,False,False,False,False,False
1810,False,False,False,False,False
1811,False,False,False,False,False
1812,False,False,False,False,False


In [ ]:
# Visualizamos la identificación de faltantes en las columnas con faltantes
encuesta[["peso","talla"]].isnull()

,peso,talla
0,False,False
1,False,False
2,False,False
3,False,False
4,False,False
...,...,...
1809,False,False
1810,False,False
1811,False,False
1812,False,False


Sin embargo, esta función nos permite calcular conteos de datos faltantes de un modo sencillo, acoplando la función `sum()`.


In [ ]:
# Conteo de datos faltantes en el dataframe, por columnas
encuesta.isnull().sum()

,0
edad,0
peso,37
talla,37
sexo,0
provincia,0


In [ ]:
# Conteo de datos faltantes en las variables con faltantes
encuesta[["peso","talla"]].isnull().sum()

,0
peso,37
talla,37


Podemos con esta función, incluso identificar si los registros con valores faltantes en `peso` son los mismos que los que tienen faltantes en `talla`.

In [ ]:
# identificamos en cuántas filas coinciden los faltantes en las dos variables
(encuesta['peso'].isnull() == encuesta['talla'].isnull()).sum()

np.int64(1814)

Así pues, de los 1814 registros existentes, en todos ellos coincide la situación de que el registro es faltante o no. Es decir, los datos faltantes se dan en las mismas filas.

Por contra, la función `notnull()` nos permitirá identificar y contabilizar los valores no perdidos:

In [ ]:
# Conteo de datos válidos, por columnas
encuesta.notnull().sum()

,0
edad,1814
peso,1777
talla,1777
sexo,1814
provincia,1814


Puesto que estas funciones actúan realmente como operadores de condición (proporcionan valores True/False), podemos seleccionar y filtrar con ellas la base de datos sin valores faltantes (o con valores faltantes), utilizando como siempre el operador de selección `[]`.

Como ya hemos visto que coinciden los registros faltantes para `peso` con los de `talla`, basta con utilizar uno de ellos



In [ ]:
# Base de datos filtrada con registros incompletos
encuesta[encuesta["peso"].isnull()].head()

,edad,peso,talla,sexo,provincia
26,63,NaN,NaN,MUJER,ALICANTE
40,20,NaN,NaN,HOMBRE,ALICANTE
80,21,NaN,NaN,MUJER,ALICANTE
170,48,NaN,NaN,HOMBRE,ALICANTE
171,47,NaN,NaN,HOMBRE,ALICANTE


In [ ]:
# Base de datos filtrada con registros completos
encuesta[encuesta["peso"].notnull()]

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


Si queremos depurar la base de datos excluyendo todas las filas en las que hay algún valor faltante, podemos utilizar la selección anterior, o bien directamente la función `dropna(how=any)`, sin tener que utilizar las columnas en la que hay valores faltantes.

Veamos cómo reproduce el resultado anterior.

In [ ]:
# Base de datos filtrada con registros completos
encuesta.dropna(how='any')

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


En este caso el argumento `how=all` no modificará la base de datos original, puesto que solo existen valores faltantes en dos de sus columnas: no hay filas completas con valores faltantes.

In [ ]:
# Base de datos filtrada excluyendo filas enteras perdidas
encuesta.dropna(how='all')

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


Si se diera el caso de que los valores faltantes en varias columnas están ubicados en filas distintas, cabría filtrar la base de datos con los valores completos para alguna de ellas, y no tomar en consideración los valores faltantes de las otras columnas. Esto lo resolvemos con el argumento `subset`, en el que indicamos las variables a tener en cuenta para identificar los valores perdidos. Esto nos permite plantear todas las posibilidades de filtrado de faltantes incorporando también el argumento `how`.

In [ ]:
# Filtramos los registros con datos completos en la variable 'peso'
encuesta.dropna(subset = ['peso'], how = 'any')

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


In [ ]:
# Filtramos los registros con datos completos, simultáneamente,
# en las variables 'peso' y 'talla'
encuesta.dropna(subset = ['peso', 'talla'], how = 'all')

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


Así, si los registros faltantes de `peso` no coincidieran con los de `talla`, excluiríamos todos los faltantes con:

In [ ]:
# Filtramos los registros con datos completos,
# en las variables 'peso' o 'talla'
encuesta.dropna(subset = ['peso', 'talla'], how = 'any')

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


Por último y a modo de curiosidad, pues más adelante aprenderemos funciones de imputación automática de valores perdidos, veamos cómo se pueden rellenar los datos faltantes.
Para ello vamos a utilizar la función `fillna(value=...)`, que se aplica a una columna, indicando en el argumento `value` el valor a imputar, esto es, con el que sustituir los valores faltantes. Para mantener la asignación sobre el dataframe original hay que añadir el argumento `inplace=True`, pues por defecto no se aplican los cambios sobre él; habría que asignar el cambio a un nuevo objeto.

Como ejemplo, vamos a sustituir los valores perdidos de la variable `peso` por el valor mínimo, y comprobar cómo cambia la media, realizando un duplicado del dataframe original.

In [ ]:
# Rellenamos los valores perdidos de 'peso' con su mínimo
print('La media con los datos faltantes es',encuesta["peso"].mean())
# creamos un duplicado
encuesta_fill=encuesta
# calculamos el mínimo peso
minimo = encuesta['peso'].min()
# y rellenamos los faltantes, asignando el mínimo
encuesta_fill['peso'] = encuesta_fill["peso"].fillna(value=minimo)
# comprobamos el cambio de media
print('La media con los datos faltantes rellenados es',encuesta_fill["peso"].mean())

La media con los datos faltantes es 69.29808666291503
La media con los datos faltantes rellenados es 68.53120176405733


## <font color='steelblue'>4. Tratamiento básico de datos

El *fileteado* de bases de datos, conocido como *slicing* en inglés,  implica la habilidad de extraer de manera precisa la información relevante de conjuntos de datos voluminosos. Mediante el uso de consultas y filtros, se puede obtener una vista específica de los datos que se ajuste a los criterios deseados. Dominar esta competencia implica comprender la estructura de la base de datos para obtener información de manera rápida y precisa.

Trabajamos en esta sección la identificación de índices en dataframes, y cómo gestionar la ordenación, el filtrado y la agrupación de información en base a criterios de selección por condiciones sobre variables en la base de datos.


Primero de todo vamos a cargar las librerías necesarias que vamos a utilizar en este cuaderno:

In [ ]:
# Importamos las librerías necesarias
%matplotlib inline
import numpy as np
import pandas as pd

### <font color='steelblue'>Ordenar

Algunas veces, para visualizar mejor la información, nos interesa mostrar los datos, o solo alguna columna, con un orden específico.  Para ello utilizamos la función **`sort_values()`** del módulo _Pandas_. El argumento `ascending` puede tener los valores booleanos _True_ o _False_, según queramos un orden ascendente o descendente, respectivamente.

Veamos cómo funciona, por ejemplo para identificar fácilmente al mostrar los datos, los valores más pequeños y los más altos.

In [ ]:
# Ordenamos edad de manera ascendente
encuesta["edad"].sort_values(ascending=True)

,edad
1237,14
1232,14
1216,14
1185,15
430,15
...,...
694,88
35,89
367,89
553,93


In [ ]:
# Ordenamos altura de manera descendente
encuesta["talla"].sort_values(ascending=False)

,talla
1210,201.0
407,193.5
1452,193.0
1186,192.0
1204,191.0
...,...
1323,NaN
1446,NaN
1588,NaN
1647,NaN


Al ordenar `edad`, se visualizan el valor menor y mayor. Al ordenar `altura`, se identifica el valor mayor, y los últimos registros que muestra corresponden a datos faltantes (`NaN`).

Podemos elegir dónde mostrar los valores faltantes; si los queremos al principio usamos el argumento `na_position='first'`. La visualización al final es la opción por defecto.

Al pedir orden descendente y los registros con valores faltantes al inicio, muestra al final los valores más pequeños de la variable, con lo cual podemos identificar el mínimo valor (133cm en este caso):

In [ ]:
# Ordenamos altura de manera descendente con los faltantes primero
encuesta["talla"].sort_values(ascending=False, na_position='first')

,talla
26,NaN
40,NaN
80,NaN
170,NaN
171,NaN
...,...
357,140.0
771,139.4
360,139.0
387,139.0


Podemos ordenar una base de datos completa o cualquier otra variable, en función de una o varias variables, incluyéndolas a modo de lista en el primer argumento de la función `sort_values()`. Al ordenar por varias variables, primero ordena respecto de la primera, y sigue con las restantes.

Por ejemplo, queremos mostrar la base de datos completa, ordenada en función de la variable `edad` por orden descendente.

In [ ]:
# Ordenamos la base de datos por edad de manera descendente
encuesta.sort_values(["edad"],ascending=False)

,edad,peso,talla,sexo,provincia
781,97,46.5,133.0,MUJER,ALICANTE
553,93,39.0,158.5,HOMBRE,CASTELLÓN
367,89,60.0,151.0,MUJER,ALICANTE
35,89,59.9,143.0,MUJER,ALICANTE
694,88,49.8,146.2,MUJER,ALICANTE
...,...,...,...,...,...
706,15,74.0,172.5,HOMBRE,ALICANTE
67,15,66.0,171.5,HOMBRE,ALICANTE
1232,14,48.0,158.5,MUJER,VALENCIA
1216,14,31.7,NaN,MUJER,VALENCIA


In [ ]:
# Ordenamos la base de datos por edad y talla de manera ascendente,
# y mostramos las 10 primeras filas
encuesta.sort_values(["edad", "talla"],ascending=True).head(10)

,edad,peso,talla,sexo,provincia
1232,14,48.0,158.5,MUJER,VALENCIA
1237,14,53.7,165.0,MUJER,VALENCIA
1216,14,31.7,NaN,MUJER,VALENCIA
709,15,69.3,151.0,HOMBRE,ALICANTE
1295,15,46.6,152.0,MUJER,VALENCIA
1224,15,42.7,152.6,MUJER,VALENCIA
1230,15,45.5,155.5,MUJER,VALENCIA
426,15,47.9,156.0,MUJER,CASTELLÓN
1003,15,55.7,156.0,MUJER,VALENCIA
722,15,51.8,157.5,MUJER,ALICANTE


Si queremos utilizar un orden diferente al ordenar por varias variables, en el argumento `ascending` asignamos una lista con las ordenaciones correspondientes a cada una de las variables por las que ordenamos.


Queremos ordenar la base de datos de modo ascendente respecto de `edad`, y luego por orden descendente respecto de `talla`.

In [ ]:
# Ordenamos ascendentemente por edad y descendentemente por altura.
# Mostramos las 10 primeras filas
encuesta.sort_values(["edad", "talla"],ascending=[True, False]).head(10)

,edad,peso,talla,sexo,provincia
1237,14,53.7,165.0,MUJER,VALENCIA
1232,14,48.0,158.5,MUJER,VALENCIA
1216,14,31.7,NaN,MUJER,VALENCIA
38,15,93.0,181.5,HOMBRE,ALICANTE
1051,15,62.0,180.0,HOMBRE,VALENCIA
580,15,63.8,178.5,HOMBRE,ALICANTE
583,15,63.2,178.5,HOMBRE,ALICANTE
1140,15,68.3,178.0,HOMBRE,VALENCIA
713,15,88.8,177.5,HOMBRE,ALICANTE
1135,15,65.5,177.0,HOMBRE,VALENCIA


Y por último, también podemos ordenar por el índice con la función **`sort_index()`**.

Ordenamos la base de datos 'encuesta2' ,indexada por el `dni`, respecto del índice.


In [ ]:
encuesta.sort_index().head()

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN


### <font color='steelblue'> Filtrar

Para encontrar los registros o variables que cumplen ciertas condiciones específicas, podemos echar mano de los filtros que conseguimos a través de los operadores de selección `[]`.

Si aplicamos un operador de condición a una variable, conseguimos una serie de *Pandas* con valores booleanos (_True_ o _False_) indicando qué registros cumplen la condición y cuáles no.

Para empezar, identificamos qué registros en la variable `provincia` coinciden con `ALICANTE`:

In [ ]:
# Identificamos los registros de la provincia de Alicante
encuesta["provincia"] == "ALICANTE"

,provincia
0,False
1,False
2,True
3,False
4,False
...,...
1809,False
1810,False
1811,False
1812,False


Una vez identificados los registros de interés (provincia de Alicante en nuestro ejemplo), para seleccionar los registros que satisfacen dicha condición, y filtrar con ellos la base de datos, utilizamos el operador de selección `[]`, incluyendo dentro la condición de filtrado.

In [ ]:
# Filtramos la base de datos con los registros de la provincia de Alicante
encuesta[encuesta["provincia"] == "ALICANTE"]

,edad,peso,talla,sexo,provincia
2,24,76.7,170.0,HOMBRE,ALICANTE
5,15,58.6,166.5,MUJER,ALICANTE
11,25,47.7,156.0,MUJER,ALICANTE
14,36,80.1,171.0,HOMBRE,ALICANTE
19,47,70.0,154.0,MUJER,ALICANTE
...,...,...,...,...,...
1637,57,83.5,164.0,HOMBRE,ALICANTE
1640,50,69.8,153.5,MUJER,ALICANTE
1646,64,64.0,154.0,MUJER,ALICANTE
1647,53,31.7,NaN,MUJER,ALICANTE


Por supuesto el filtro se puede referir a valores coincidentes con uno dado, pero también con valores superiores o inferiores a uno dado, para lo que utilizamos los operadores `>` o `<`.

Por ejemplo, queremos filtrar los registros con talla superior a 170cm:

In [ ]:
# Registros con talla superior a 170cm
encuesta[encuesta["talla"] > 170]

,edad,peso,talla,sexo,provincia
1,18,68.7,172.0,HOMBRE,CASTELLÓN
8,25,55.2,174.0,HOMBRE,CASTELLÓN
9,34,76.0,181.5,HOMBRE,CASTELLÓN
14,36,80.1,171.0,HOMBRE,ALICANTE
15,45,88.9,171.0,HOMBRE,CASTELLÓN
...,...,...,...,...,...
1722,67,84.0,177.5,HOMBRE,VALENCIA
1725,72,88.1,182.1,HOMBRE,VALENCIA
1726,68,80.9,170.5,HOMBRE,VALENCIA
1736,76,91.1,172.0,HOMBRE,VALENCIA


Una vez hecho el filtrado, podemos seleccionar o simplemente visualizar cualquiera de las variables individualizadas de la base de datos.


Por ejemplo, queremos mostrar solo los valores de `provincia` para los registros de sujetos más altos (talla>170cm):

In [ ]:
# Valor de 'provincia' para los sujetos más altos (talla>170cm)
encuesta[encuesta["talla"] > 170]["provincia"]

,provincia
1,CASTELLÓN
8,CASTELLÓN
9,CASTELLÓN
14,ALICANTE
15,CASTELLÓN
...,...
1722,VALENCIA
1725,VALENCIA
1726,VALENCIA
1736,VALENCIA


Si queremos recuperar en un objeto el número de registros que cumplen esta condición, podemos utilizar la función `count()` a continuación.

In [ ]:
# Nº de registros de sujetos más altos (talla>170cm)
encuesta[encuesta["talla"] > 170]['provincia'].count()

np.int64(439)

La selección, como ya vimos, podemos hacerla también con la función `loc[]`, indicando primero la condición y luego, seguida de una coma, la variable o variables a considerar/visualizar.

Repetimos el filtrado anterior con `loc`.

In [ ]:
# Valor de 'provincia' para los sujetos más altos (talla>170cm)
encuesta.loc[encuesta["talla"] > 170, 'provincia']

,provincia
1,CASTELLÓN
8,CASTELLÓN
9,CASTELLÓN
14,ALICANTE
15,CASTELLÓN
...,...
1722,VALENCIA
1725,VALENCIA
1726,VALENCIA
1736,VALENCIA


Para seleccionar más de una variable, utilizamos una lista con las variables de interés.

In [ ]:
# Valores de 'provincia' y 'peso' para los sujetos más altos (talla>170cm)
encuesta.loc[encuesta["talla"] > 170, ["provincia", "peso"]]

,provincia,peso
1,CASTELLÓN,68.7
8,CASTELLÓN,55.2
9,CASTELLÓN,76.0
14,ALICANTE,80.1
15,CASTELLÓN,88.9
...,...,...
1722,VALENCIA,84.0
1725,VALENCIA,88.1
1726,VALENCIA,80.9
1736,VALENCIA,91.1


Además, podemos combinar más de una condición utilizando los operadores `&` y `|` para las condiciones `y` u `o`, respectivamente, e incluyendo entre paréntesis cada una de las condiciones a verificar.

Por ejemplo, queremos visualizar los datos de los registros:
1.  en la provincia de Alicante y con altura superior a 170cm
2. con edad inferior a 25 o superior a 60 años.

In [ ]:
# 1. Registros de la provincia de Alicante y altura superior a 170cm
encuesta[(encuesta["provincia"] == "ALICANTE") & (encuesta["talla"] > 170)]

,edad,peso,talla,sexo,provincia
14,36,80.1,171.0,HOMBRE,ALICANTE
38,15,93.0,181.5,HOMBRE,ALICANTE
42,22,65.1,172.0,HOMBRE,ALICANTE
43,15,89.5,171.5,HOMBRE,ALICANTE
44,17,62.2,172.0,HOMBRE,ALICANTE
...,...,...,...,...,...
912,43,89.3,171.5,HOMBRE,ALICANTE
919,60,103.3,174.0,HOMBRE,ALICANTE
926,65,89.5,172.5,HOMBRE,ALICANTE
1153,18,71.6,176.0,HOMBRE,ALICANTE


De las encuestas que se han pasado, hay 134 sujetos que cumplen esta condición.

Si queremos seleccionar los sujetos que tienen menos de 25 años o más de 60:

In [ ]:
# 2. Registros con edad inferior a 25 o superior a 60 años.

# **Nota**: Cuidado si hemos cambiado  la variable `edad` a tipo _string_, en cuyo caso debemos  indicar el valor entrecomillado, `edad < "25"` y `edad > "60"`.

encuesta[(encuesta["edad"] < "25") | (encuesta["edad"] > "60")]

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


o transfromando la variable a float

In [ ]:
encuesta[(encuesta["edad"].astype(float) < 25) | (encuesta["edad"].astype(float) > 60)]

,edad,peso,talla,sexo,provincia
0,18,66.4,163.5,HOMBRE,CASTELLÓN
1,18,68.7,172.0,HOMBRE,CASTELLÓN
2,24,76.7,170.0,HOMBRE,ALICANTE
3,17,55.6,160.0,MUJER,CASTELLÓN
4,16,46.6,163.5,MUJER,CASTELLÓN
...,...,...,...,...,...
1809,71,80.2,157.0,MUJER,VALENCIA
1810,81,68.3,150.0,MUJER,VALENCIA
1811,86,66.5,154.5,MUJER,VALENCIA
1812,67,74.1,150.0,MUJER,VALENCIA


Convertimos la variable a float

In [ ]:
encuesta["edad"] = encuesta["edad"].astype(float)

Por supuesto, podemos combinar los filtros con la ordenación, y así ordenar los registros previamente filtrados de una base de datos.


Queremos ordenar por la variable `edad` y de modo ascendente, los registros de 'encuesta' que tienen edades inferiores a 25 o superiores a 60 años:

In [ ]:
encuesta[(encuesta["edad"] < 25) | (encuesta["edad"] > 60)].sort_values(["edad"])

,edad,peso,talla,sexo,provincia
1216,14.0,31.7,NaN,MUJER,VALENCIA
1237,14.0,53.7,165.0,MUJER,VALENCIA
1232,14.0,48.0,158.5,MUJER,VALENCIA
45,15.0,67.0,170.0,HOMBRE,ALICANTE
1123,15.0,43.0,161.8,HOMBRE,VALENCIA
...,...,...,...,...,...
694,88.0,49.8,146.2,MUJER,ALICANTE
35,89.0,59.9,143.0,MUJER,ALICANTE
367,89.0,60.0,151.0,MUJER,ALICANTE
553,93.0,39.0,158.5,HOMBRE,CASTELLÓN


Cuando queremos identificar exclusivamente ciertos valores específicos de alguna variable, podemos proceder incluyendo cada una de las condiciones de igualdad precisas, o de modo simplificado a través de la función **`isin()`**, que pregunta (establece una condición) sobre si los valores están en el conjunto de datos suministrado como lista en su argumento.

Veamos cómo seleccionar/filtrar la información de los sujetos que tienen 21 o 25 o 30 años:

In [ ]:
# Selección de valores específicos con condiciones individualizadas
encuesta[(encuesta["edad"] == 21) | (encuesta["edad"] == 25) |
            (encuesta["edad"] == 30)]

,edad,peso,talla,sexo,provincia
8,25.0,55.2,174.0,HOMBRE,CASTELLÓN
11,25.0,47.7,156.0,MUJER,ALICANTE
12,25.0,65.4,157.5,MUJER,CASTELLÓN
52,21.0,65.0,173.0,HOMBRE,ALICANTE
55,21.0,65.2,176.0,HOMBRE,ALICANTE
...,...,...,...,...,...
1422,25.0,68.5,169.5,MUJER,VALENCIA
1423,30.0,58.9,161.0,MUJER,VALENCIA
1428,30.0,64.3,163.0,MUJER,VALENCIA
1429,25.0,62.9,162.0,MUJER,VALENCIA


In [ ]:
# Selección de valores específicos con 'isin()'
encuesta[encuesta["edad"].isin([21, 25, 30])]

,edad,peso,talla,sexo,provincia
8,25.0,55.2,174.0,HOMBRE,CASTELLÓN
11,25.0,47.7,156.0,MUJER,ALICANTE
12,25.0,65.4,157.5,MUJER,CASTELLÓN
52,21.0,65.0,173.0,HOMBRE,ALICANTE
55,21.0,65.2,176.0,HOMBRE,ALICANTE
...,...,...,...,...,...
1422,25.0,68.5,169.5,MUJER,VALENCIA
1423,30.0,58.9,161.0,MUJER,VALENCIA
1428,30.0,64.3,163.0,MUJER,VALENCIA
1429,25.0,62.9,162.0,MUJER,VALENCIA


### <font color='steelblue'>Agrupar

Con frecuencia nos interesará agrupar los datos disponibles según los diferentes niveles de ciertas variables cualitativas. Estas agrupaciones son útiles por ejemplo, para calcular medias u otros descriptivos por grupos, y tener así una medida de comparación entre ellos.

La agrupación de una base de datos según alguna variable cualitativa la resolvemos con la función **`groupby()`** e incluyendo como argumento el nombre de la variable clasificatoria o de agrupación. Si a continuación nos interesan solo algunas variables de la base de datos, las seleccionamos entrecomilladas. Y si finalmente queremos presentar algún descriptivo sobre ellas, lo aplicamos a continuación con la función correspondiente.
```
# Calcula la media en las variables var1 y var2, para cada uno de los
# grupos que genera el cruce de las variables var_g1 y var_g2:
dataframe.groupby(["var_g1","var_g2"])[["var1","var2"]].mean()
```


Queremos por ejemplo, calcular la talla media diferenciando por provincia, y luego calcular la talla y la edad medias, diferenciando por provincia y por sexo:

In [ ]:
# Talla media en cada una de las provincias
encuesta.groupby("provincia")["talla"].mean()

,talla
provincia,
ALICANTE,162.261755
CASTELLÓN,165.200000
VALENCIA,164.549618


In [ ]:
# Talla y edad medias en cada provincia, diferenciando por sexo
encuesta.groupby(["provincia", "sexo"])[["talla","edad"]].mean()

talla       edad
provincia sexo                         
ALICANTE  HOMBRE  169.445725  42.964413
          MUJER   157.024661  45.361979
CASTELLÓN HOMBRE  171.933628  38.964912
          MUJER   158.219266  40.872727
VALENCIA  HOMBRE  171.697156  39.273364
          MUJER   158.456162  41.507042

Este tipo de descriptivos nos permite empezar a analizar cómo es la base de datos. Así por ejemplo, observamos que los hombres más altos (en media) se encuentran en Castellón, y las mujeres más altas (en media) en Valencia.

Si no seleccionamos ninguna variable en particular sobre la base de datos, al realizar la agrupación y solicitar un descriptivo cualquiera, lo aplicará a todas las variables susceptibles de proporcionar dicho descriptivo. Si se trata por ejemplo de una media, se emplearán solo las variables cuantitativas (enteras o decimales). Si se trata de un conteo, se aplicará a todas las variables.

In [ ]:
# promedio de las variables cuantitativas clasificadas por provincia
encuesta.groupby(["provincia"]).mean(numeric_only=True)

,edad,peso,talla
provincia,,,
ALICANTE,44.348872,67.541053,162.261755
CASTELLÓN,39.901786,70.093750,165.200000
VALENCIA,40.473514,68.864649,164.549618


In [ ]:
# conteo de todas las variables clasificadas por provincia
encuesta.groupby("provincia").count()

,edad,peso,talla,sexo
provincia,,,,
ALICANTE,665,665,638,665
CASTELLÓN,224,224,222,224
VALENCIA,925,925,917,925


In [ ]:
# promedio de las variables cuantitativas clasificadas por provincia y sexo
encuesta.groupby(["provincia","sexo"]).mean(numeric_only=True)

edad       peso       talla
provincia sexo                                    
ALICANTE  HOMBRE  42.964413  73.211032  169.445725
          MUJER   45.361979  63.391927  157.024661
CASTELLÓN HOMBRE  38.964912  76.822807  171.933628
          MUJER   40.872727  63.120000  158.219266
VALENCIA  HOMBRE  39.273364  75.293692  171.697156
          MUJER   41.507042  63.328169  158.456162

## <font color='steelblue'> 5. Crear y codificar variables

La creación de nuevas variables es necesaria en ocasiones para identificar patrones y tendencias en conjuntos de datos existentes y derivar información adicional que pueda ser útil para el análisis. Se pueden generar variables que aporten valor adicional a los modelos o estudios realizados, permitiendo obtener una visión más completa y enriquecida de los datos, ya sea a base de combinar información en diversas variables, bien a base de categorizar o incluso recodificar variables ya existentes. Estas son las premisas que estudiamos en esta sección.


En el cuaderno anterior vimos cómo crear variables nuevas asignando valores con una lista (`data_frame["nueva_variable"] = [1, 2, 3, 4]`) o utilizando operadores sencillos sobre algunas ya existentes (`data_frame["nueva_variable"] = data_frame["variable1"] + data_frame["variable2"]`).

Si la base de datos con la que estamos trabajando tiene muchos registros (filas), es difícil crear una variable pasando una lista. Además, si los cálculos implican a más de dos o tres variables, e incluso cálculos complejos, trabajar con los operadores convencionales y la asignación del resultado en una nueva columna puede ser un método un tanto farragoso. Por todo ello se propone una nueva estrategia consistente. en programar funciones específicas que calculan y aplican los valores correspondientes para crear nuevas columnas en el dataframe.


A modo de ejemplo vamos a crear la función `bmi`, que calcula el *índice de masa corporal* (BMI) a partir de la `talla` y el `peso` en el dataframe *encuesta*. Puesto que involucra cálculos con más de una variable de la base de datos, necesitamos de otra función que la aplique a las variables de dicha base de datos; esta función es **`vectorize()`** de _NumPy_. En el argumento de `vectorize` indicamos la función a aplicar y fuera, entre paréntesis `("var1","var2",...)`, las variables a las que se aplica.

Para poder calcular este índice, puesto que la talla viene dada en centímetros, hemos de convertirlos a metros:

$$bmi = \frac{peso (kg)}{talla (m) ^2}.$$

In [ ]:
# Determinación del bmi

# Primero convertimos la 'talla' a metros
encuesta["talla"] = encuesta["talla"]/100

# Definimos la función 'bmi'
def bmi(x, y):
  '''
  Calcula la división del primer argumento entre el cuadrado del segundo

  Argumentos:
  - x: objeto de Python que sea operable
  - y: objeto de Python que sea operable

  Devuelve:
    el valor de 'x' entre 'y' al cuadrado, redondeado a las centésimas
  '''
  return round(x/y**2, 2)

# Aplicamos la función 'bmi' a las variables 'peso' y 'talla', para crear la variable 'bmi'
encuesta["bmi"] = np.vectorize(bmi)(encuesta["peso"], encuesta["talla"])

# Mostramos las 5 primeras filas
encuesta.head()

,edad,peso,talla,sexo,provincia,bmi
0,18.0,66.4,1.635,HOMBRE,CASTELLÓN,24.84
1,18.0,68.7,1.720,HOMBRE,CASTELLÓN,23.22
2,24.0,76.7,1.700,HOMBRE,ALICANTE,26.54
3,17.0,55.6,1.600,MUJER,CASTELLÓN,21.72
4,16.0,46.6,1.635,MUJER,CASTELLÓN,17.43


La función **`vectorize()`** se utiliza cuando la función se aplica a más de una  variable de la base de datos. Si queremos construir una nueva variable aplicando una función a una única variable, utilizamos la función **`apply()`** de _Pandas_.

Para ejemplificar su uso, una vez construida la variable `bmi`, queremos utilizarla para clasificar a los sujetos en la categorización que propone la OMS:

<center><img src=https://raw.githubusercontent.com/ia4legos/Statistics/main/images/img_170_20_10.png width="350" height="200"></center>

Definimos a continuación la función `clasificar_bmi()`, para clasificar a los sujetos según su `bmi` en dichos niveles, y guardamos el resultado en una nueva variable llamada `bmi_agrupado`.

In [ ]:
# Categorización de variables numéricas con función propia
# clasificación OMS de los sujetos por su bmi
def clasificar_bmi(x):
  '''
  Clasifica a los individuos en grupos dependiendo de su bmi

  Argumentos:
    - x: número

  Devuelve:
    - si el valor de 'x' es menor que 18.5, Infrapeso
    - si el valor de 'x' está entre 18.5 (inclusive) y 25, Normal
    - si el valor de 'x' está entre 25 (inclusice) y 30, Sobrepeso
    - si el valor de 'x' está entre 30 (inclusive) y 35, Obesidad1
    - si el valor de 'x' está entre 35 (inclusive) y 40, Obesidad2
    - si el valor de 'x' es mayor o igual a 40 (o en caso contrario), Obesidad3
  '''
  if x<18.5:
    return "Infrapeso"
  elif 18.5<=x<25:
    return "Normal"
  elif 25<=x<30:
    return "Sobrepeso"
  elif 30<=x<35:
    return "Obesidad1"
  elif 35<=x<40:
    return "Obesidad2"
  else:
    return "Obesidad3"

encuesta["bmi_agrupado"] = encuesta["bmi"].apply(clasificar_bmi)
encuesta.head()

,edad,peso,talla,sexo,provincia,bmi,bmi_agrupado
0,18.0,66.4,1.635,HOMBRE,CASTELLÓN,24.84,Normal
1,18.0,68.7,1.720,HOMBRE,CASTELLÓN,23.22,Normal
2,24.0,76.7,1.700,HOMBRE,ALICANTE,26.54,Sobrepeso
3,17.0,55.6,1.600,MUJER,CASTELLÓN,21.72,Normal
4,16.0,46.6,1.635,MUJER,CASTELLÓN,17.43,Infrapeso


Hemos hecho lo que se denomina **categorización de una variable numérica por niveles de respuesta**. Esto lo podemos hacer automáticamente con la función **`cut()`** de _Pandas_, sin necesidad de programar una nueva función. Los argumentos de la función `cut()` son: la variable a categorizar, los límites de cada nivel (`bins`), las etiquetas de cada nivel (`labels`). Si queremos que los intervalos de respuesta estén abiertos por la derecha, hemos de indicar en los argumentos `right=False`, pues por defecto crea intervalos cerrados por la derecha (`límite inferior < x <= límite superior`) y nos interesa lo contrario (`límite inferior <= x < límite superior`).

Para aplicar la función `cut()` eliminamos la que hemos creado antes con la categorización, `bmi_agrupado` (con `drop()`), y utilizamos el mismo nombre para volcar el resultado.

In [ ]:
# Categorización de variables numéricas con cut()

# eliminamos la variable 'bmi_agrupado' para un nuevo dataframe
encuesta2 = encuesta.drop(columns='bmi_agrupado')

# volvemos a crear 'bmi_agrupado' con cut()
encuesta2["bmi_agrupado"] = pd.cut(encuesta2["bmi"],
                                   bins = [0, 18.5, 25, 30, 35, 40, encuesta2["bmi"].max()],
                                   labels = ["Infrapeso", "Normal", "Sobrepeso", "Obesidad1", "Obesidad2", "Obesidad3"],
                                   right=False)
encuesta2.head()

,edad,peso,talla,sexo,provincia,bmi,bmi_agrupado
0,18.0,66.4,1.635,HOMBRE,CASTELLÓN,24.84,Normal
1,18.0,68.7,1.720,HOMBRE,CASTELLÓN,23.22,Normal
2,24.0,76.7,1.700,HOMBRE,ALICANTE,26.54,Sobrepeso
3,17.0,55.6,1.600,MUJER,CASTELLÓN,21.72,Normal
4,16.0,46.6,1.635,MUJER,CASTELLÓN,17.43,Infrapeso


Sobre una variable categórica, en determinados contextos nos puede interesar recodificar los valores de respuesta, esto es, asignar otras etiquetas a los niveles de respuesta. Esto se llama **recodificación de variables categóricas**. Para esta tarea podemos utilizar varias funciones, como **`replace()`** o **`map()`**.
```
dataframe.replace({'var1':'var1_nueva','var2':'var2_nueva',...})
```

Seguimos sobre el mismo ejemplo. Dado que la variable `bmi_agrupado` tiene varios niveles de Obesidad (1,2,3), si queremos agruparlos todos ellos en un único nivel, esto es, **colapsarlos**, aplicamos los cambios de etiquetas (Obesidad1=Obesidad, Obesidad2=Obesidad y Obesidad3=Obesidad) a través de la función `replace()` y un diccionario con los cambios. Aplicamos los cambios sobre la misma variable `bmi_agrupado` con el argumento `inplace=True`.


In [ ]:
# Recodificar una variable: colapsar niveles

# los niveles de respuesta original
print(encuesta.bmi_agrupado)
# colapsamos todos los niveles de Obesidad1,Obesidad2 y Obesidad3 en uno solo: Obesidad
encuesta.replace({"Obesidad1": "Obesidad", "Obesidad2": "Obesidad", "Obesidad3": "Obesidad"}, inplace=True)
print(encuesta.bmi_agrupado)

0          Normal
1          Normal
2       Sobrepeso
3          Normal
4       Infrapeso
          ...    
1809    Obesidad1
1810    Obesidad1
1811    Sobrepeso
1812    Obesidad1
1813    Obesidad2
Name: bmi_agrupado, Length: 1814, dtype: object
0          Normal
1          Normal
2       Sobrepeso
3          Normal
4       Infrapeso
          ...    
1809     Obesidad
1810     Obesidad
1811    Sobrepeso
1812     Obesidad
1813     Obesidad
Name: bmi_agrupado, Length: 1814, dtype: object


Por supuesto, con el mismo método podemos realizar cualquier otra recodificación, como la de cambiar por números las etiquetas de los niveles en una variable.

Por ejemplo, en la variable `provincia` podemos sustituir Alicante por un 1, Castellón por un 2 y Valencia por un 3.

In [ ]:
# Recodificación de variables: variables nominales a codificación numérica
pd.set_option('future.no_silent_downcasting', True)
encuesta2.replace({"ALICANTE": 1, "CASTELLÓN": 2, "VALENCIA": 3})

,edad,peso,talla,sexo,provincia,bmi,bmi_agrupado
0,18.0,66.4,1.635,HOMBRE,2,24.84,Normal
1,18.0,68.7,1.720,HOMBRE,2,23.22,Normal
2,24.0,76.7,1.700,HOMBRE,1,26.54,Sobrepeso
3,17.0,55.6,1.600,MUJER,2,21.72,Normal
4,16.0,46.6,1.635,MUJER,2,17.43,Infrapeso
...,...,...,...,...,...,...,...
1809,71.0,80.2,1.570,MUJER,3,32.54,Obesidad1
1810,81.0,68.3,1.500,MUJER,3,30.36,Obesidad1
1811,86.0,66.5,1.545,MUJER,3,27.86,Sobrepeso
1812,67.0,74.1,1.500,MUJER,3,32.93,Obesidad1


Con la función `map()` podemos realizar la misma recodificación, también a través de un diccionario con los cambios.

In [ ]:
# Recodificación de variables: variables nominales a codificación numérica
encuesta2["provincia"].map({"ALICANTE": 1, "CASTELLÓN": 2, "VALENCIA": 3})

,provincia
0,2
1,2
2,1
3,2
4,2
...,...
1809,3
1810,3
1811,3
1812,3


Para que los cambios se apliquen, hemos de asignar el resultado de la recodificación en la mismo u otra columna. Optamos por asignarlo en otra columna nueva llamada `codigo_provincia`.

In [ ]:
# Creación de nueva variable con la recodificación de provincia
encuesta2["codigo_provincia"] = encuesta2["provincia"].map({"ALICANTE": 1, "CASTELLÓN": 2, "VALENCIA": 3})
encuesta2.head()

,edad,peso,talla,sexo,provincia,bmi,bmi_agrupado,codigo_provincia
0,18.0,66.4,1.635,HOMBRE,CASTELLÓN,24.84,Normal,2
1,18.0,68.7,1.720,HOMBRE,CASTELLÓN,23.22,Normal,2
2,24.0,76.7,1.700,HOMBRE,ALICANTE,26.54,Sobrepeso,1
3,17.0,55.6,1.600,MUJER,CASTELLÓN,21.72,Normal,2
4,16.0,46.6,1.635,MUJER,CASTELLÓN,17.43,Infrapeso,2


Si en un momento determinado queremos prescindir de una de las dos columnas que contienen la misma información, `provincia` y `codigo_provincia`, basta eliminarla con la función `drop()`.


In [ ]:
# Eliminamos la variable codigo_provincia por tener la misma información que provincia
encuesta2.drop(columns="codigo_provincia", inplace=True)
encuesta2.head()

,edad,peso,talla,sexo,provincia,bmi,bmi_agrupado
0,18.0,66.4,1.635,HOMBRE,CASTELLÓN,24.84,Normal
1,18.0,68.7,1.720,HOMBRE,CASTELLÓN,23.22,Normal
2,24.0,76.7,1.700,HOMBRE,ALICANTE,26.54,Sobrepeso
3,17.0,55.6,1.600,MUJER,CASTELLÓN,21.72,Normal
4,16.0,46.6,1.635,MUJER,CASTELLÓN,17.43,Infrapeso
